File: tests/test_truth_gate_policy.py
import json
import re
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


class TruthGatePolicyTests(unittest.TestCase):
    def test_structured_authorities_and_reader_facing_projection_policy(self) -> None:
        architecture = (ROOT / "docs" / "ARCHITECTURE.md").read_text(encoding="utf-8")
        package = json.loads((ROOT / "package.json").read_text(encoding="utf-8"))
        acceptance = json.loads(
            (ROOT / "conformance" / "native-host-acceptance.json").read_text(encoding="utf-8")
        )
        capabilities = json.loads(
            (ROOT / "core" / "hakim-skill" / "capabilities.json").read_text(encoding="utf-8")
        )
        version = (ROOT / "core" / "hakim-skill" / "VERSION").read_text(encoding="utf-8").strip()
        history_dir = ROOT / "conformance" / "history"

        self.assertIn("## Truth-gate policy", architecture)
        self.assertIn("Structured facts have structured authorities", architecture)
        self.assertRegex(architecture, re.compile(r"machine-readable or structural sources", re.I))
        self.assertRegex(architecture, re.compile(r"prose order is not a product invariant", re.I))
        self.assertRegex(
            architecture,
            re.compile(r"structured authority.*focused test", re.I | re.S),
        )

        # Product identity and current acceptance are structural authorities.
        self.assertEqual(package["version"], version)
        self.assertEqual(acceptance["product_version"], version)
        self.assertEqual(acceptance["schema_version"], 1)
        self.assertIn(acceptance["overall_status"], {"PASS", "HOLD_FOR_LIVE_HOST_EVIDENCE"})
        self.assertEqual(capabilities["schema_version"], 1)
        self.assertTrue(capabilities["capabilities"])

        # Historical evidence has its own retained structural surface rather than
        # being inferred from a sentence in current documentation.
        self.assertTrue(history_dir.is_dir())
        history_files = sorted(history_dir.glob("native-host-acceptance-*.json"))
        self.assertTrue(history_files, "historical acceptance authority must remain populated")
        for path in history_files:
            historical = json.loads(path.read_text(encoding="utf-8"))
            self.assertEqual(historical["schema_version"], 1)
            self.assertNotEqual(historical["product_version"], "")

        for retired in (
            "scripts/check_metadata_truth_consistency.mjs",
            "scripts/check_current_truth_consistency.mjs",
        ):
            self.assertFalse((ROOT / retired).exists(), f"retired prose-oracle returned: {retired}")

        first_run = (ROOT / "tests" / "test_public_first_run_contract.mjs").read_text(encoding="utf-8")
        self.assertIsNone(
            re.search(r"README[^\n]{0,120}[a-f0-9]{40}", first_run, re.I),
            "first-run gate must not require a hardcoded commit SHA inside README prose",
        )

        # The public-doc gate may enforce reader-facing structure and reject stale
        # operational surfaces, but it must not copy-lock one status sentence or
        # milestone ordering into the product contract.
        self.assertIsNone(
            re.search(r"readme\.includes\(['\"]Hakim `['\"]\s*\+\s*version", first_run),
            "first-run gate must not copy-lock README status prose",
        )
        self.assertNotIn("P0 before F05", first_run)
        self.assertIn("retiredOperationalDocs", first_run)


if __name__ == "__main__":
    unittest.main()

File: tests/test_truth_gate_policy.mjs
#!/usr/bin/env node
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const read = (relative) => fs.readFileSync(path.join(root, relative), 'utf8');
const architecture = read('docs/ARCHITECTURE.md');
const packageJson = JSON.parse(read('package.json'));
const acceptance = JSON.parse(read('conformance/native-host-acceptance.json'));
const capabilities = JSON.parse(read('core/hakim-skill/capabilities.json'));

assert.match(architecture, /^## Truth-gate policy$/m);
assert.match(architecture, /Structured facts have structured authorities/);
assert.match(architecture, /negative tripwires/i);
assert.match(architecture, /not semantic proof/i);
assert.match(architecture, /cannot promote acceptance or release state/i);
assert.match(architecture, /prefer adding or reusing a structured authority/i);

// Machine facts used by public truth checks must actually exist structurally.
assert.equal(packageJson.version, read('core/hakim-skill/VERSION').trim());
assert.equal(acceptance.product_version, packageJson.version);
assert.equal(acceptance.schema_version, 1);
assert.ok(['PASS', 'HOLD_FOR_LIVE_HOST_EVIDENCE'].includes(acceptance.overall_status));
assert.equal(capabilities.schema_version, 1);
assert.ok(Array.isArray(capabilities.capabilities) && capabilities.capabilities.length > 0);

// F-4 regression: do not reintroduce a positive README prose/commit-SHA oracle.
for (const candidate of [
  'scripts/check_metadata_truth_consistency.mjs',
  'scripts/check_current_truth_consistency.mjs',
]) {
  assert.equal(fs.existsSync(path.join(root, candidate)), false, `${candidate} must remain retired`);
}
const firstRun = read('tests/test_public_first_run_contract.mjs');
assert.ok(!/README[^\n]{0,120}[a-f0-9]{40}/i.test(firstRun), 'first-run gate must not require a hardcoded commit SHA inside README prose');

console.log('truth-gate policy OK: structured authorities govern facts; prose checks are negative tripwires only');
